File: edge/app/backup/filters.py
from __future__ import annotations

import fnmatch
import os
import stat
from dataclasses import dataclass
from pathlib import Path, PurePosixPath

from app.backup.discovery import JobDefinition, UPLOAD_DIR_FILENAME


@dataclass(slots=True)
class DiscoveredFile:
    source_path: Path
    archive_path: str
    size: int
    mtime_ns: int


def build_file_list(job: JobDefinition, logger) -> list[DiscoveredFile]:
    files: list[DiscoveredFile] = []
    stack = [job.root_path]
    visited_dirs: set[Path] = set()

    while stack:
        current_dir = stack.pop()
        visit_key = current_dir.resolve(strict=False) if job.follow_symlinks else current_dir
        if visit_key in visited_dirs:
            continue
        visited_dirs.add(visit_key)

        try:
            entries = list(os.scandir(current_dir))
        except (FileNotFoundError, NotADirectoryError, PermissionError, OSError) as exc:
            logger.warning("skipped_missing path=%s detail=%s", current_dir, exc)
            continue

        for entry in entries:
            relative_path = Path(entry.path).relative_to(job.root_path).as_posix()
            if relative_path == UPLOAD_DIR_FILENAME:
                continue
            if not job.include_hidden and _contains_hidden(relative_path):
                continue
            if _matches_exclude(relative_path, job.exclude_patterns):
                continue

            try:
                if entry.is_dir(follow_symlinks=job.follow_symlinks):
                    stack.append(Path(entry.path))
                    continue
                stat_result = entry.stat(follow_symlinks=job.follow_symlinks)
            except (FileNotFoundError, PermissionError, OSError) as exc:
                logger.warning("skipped_missing path=%s detail=%s", entry.path, exc)
                continue

            if not stat.S_ISREG(stat_result.st_mode):
                continue

            files.append(
                DiscoveredFile(
                    source_path=Path(entry.path),
                    archive_path=relative_path,
                    size=stat_result.st_size,
                    mtime_ns=stat_result.st_mtime_ns,
                )
            )

    files.sort(key=lambda item: item.archive_path)
    return files


def _contains_hidden(relative_path: str) -> bool:
    return any(part.startswith(".") for part in PurePosixPath(relative_path).parts)


def _matches_exclude(relative_path: str, patterns: list[str]) -> bool:
    basename = PurePosixPath(relative_path).name

    for pattern in patterns:
        normalized = pattern.strip()
        if not normalized:
            continue

        if normalized.endswith("/"):
            prefix = normalized.rstrip("/")
            if (
                relative_path == prefix
                or relative_path.startswith(f"{prefix}/")
                or f"/{prefix}/" in relative_path
            ):
                return True
            continue

        if "/" in normalized:
            if fnmatch.fnmatch(relative_path, normalized):
                return True
            continue

        if fnmatch.fnmatch(relative_path, normalized) or fnmatch.fnmatch(basename, normalized):
            return True

    return False

File: edge/internal/backup/filters.go
package backup

import (
	"cmp"
	"os"
	"path/filepath"
	"slices"
	"strings"
)

// DiscoveredFile describes a single file collected during a job scan.
type DiscoveredFile struct {
	SourcePath  string
	ArchivePath string // slash-separated path relative to job root
	Size        int64
	MtimeNs     int64
}

// BuildFileList walks the job's root directory and returns regular files, applying
// exclude patterns and hidden-file filtering as configured in the job.
func BuildFileList(job *JobDefinition, warnf func(string, ...interface{})) ([]*DiscoveredFile, error) {
	var files []*DiscoveredFile
	stack := []string{job.RootPath}
	visited := make(map[string]bool)

	for len(stack) > 0 {
		curDir := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		visitKey := curDir
		if job.FollowSymlinks {
			if abs, err := filepath.EvalSymlinks(curDir); err == nil {
				visitKey = abs
			}
		}
		if visited[visitKey] {
			continue
		}
		visited[visitKey] = true

		entries, err := os.ReadDir(curDir)
		if err != nil {
			if warnf != nil {
				warnf("skipped_missing path=%s detail=%s", curDir, err)
			}
			continue
		}

		for _, entry := range entries {
			entryPath := filepath.Join(curDir, entry.Name())
			rel, err := filepath.Rel(job.RootPath, entryPath)
			if err != nil {
				continue
			}
			archivePath := filepath.ToSlash(rel)

			if archivePath == UploadDirFilename {
				continue
			}
			if !job.IncludeHidden && containsHidden(archivePath) {
				continue
			}
			if matchesExclude(archivePath, job.ExcludePatterns) {
				continue
			}

			info, err := entry.Info()
			if err != nil {
				if warnf != nil {
					warnf("skipped_missing path=%s detail=%s", entryPath, err)
				}
				continue
			}

			if entry.Type()&os.ModeSymlink != 0 {
				if job.FollowSymlinks {
					resolved, err := filepath.EvalSymlinks(entryPath)
					if err != nil {
						if warnf != nil {
							warnf("skipped_missing path=%s detail=%s", entryPath, err)
						}
						continue
					}
					info, err = os.Stat(resolved)
					if err != nil {
						if warnf != nil {
							warnf("skipped_missing path=%s detail=%s", entryPath, err)
						}
						continue
					}
					if info.IsDir() {
						stack = append(stack, entryPath)
						continue
					}
				} else {
					continue
				}
			} else if info.IsDir() {
				stack = append(stack, entryPath)
				continue
			}

			if !info.Mode().IsRegular() {
				continue
			}

			files = append(files, &DiscoveredFile{
				SourcePath:  entryPath,
				ArchivePath: archivePath,
				Size:        info.Size(),
				MtimeNs:     info.ModTime().UnixNano(),
			})
		}
	}

	slices.SortFunc(files, func(a, b *DiscoveredFile) int {
		return cmp.Compare(a.ArchivePath, b.ArchivePath)
	})
	return files, nil
}

func containsHidden(archivePath string) bool {
	for _, part := range strings.Split(archivePath, "/") {
		if strings.HasPrefix(part, ".") {
			return true
		}
	}
	return false
}

func matchesExclude(archivePath string, patterns []string) bool {
	basename := filepath.Base(archivePath)

	for _, pattern := range patterns {
		normalized := strings.TrimSpace(pattern)
		if normalized == "" {
			continue
		}

		if strings.HasSuffix(normalized, "/") {
			prefix := strings.TrimSuffix(normalized, "/")
			if archivePath == prefix ||
				strings.HasPrefix(archivePath, prefix+"/") ||
				strings.Contains(archivePath, "/"+prefix+"/") {
				return true
			}
			continue
		}

		if strings.Contains(normalized, "/") {
			if matched, _ := filepath.Match(normalized, archivePath); matched {
				return true
			}
			continue
		}

		if matched, _ := filepath.Match(normalized, archivePath); matched {
			return true
		}
		if matched, _ := filepath.Match(normalized, basename); matched {
			return true
		}
	}
	return false
}
